[NU-1772] Scenario activities BE implementation

mgoworko opened the pull request:
## Describe your changes

What is done:
- new API contract for FE (in PR https://github.com/TouK/nussknacker/pull/6822)
- BE implemenetation of that API in this PR
- new table that replaces actions and comments
- possibility to save and return to FE all scenario-related activities
- full backwards compatibility with old FE API, and old Action mechanism, only changed the db underneath
- migration of data from 2 old tables - comments and actions - to new "scenario_activities" table
- some integration tests of the new API


What is left to do in further PRs
1. Veryfing what is already done and saving information of scenario activities that has not been saved until now (like incoming and outgoing migrations, automatic updates of scenarios etc.)
2. Integrating periodics into this new scenario activity mechanism.
3. Clean up endpoint with scenario details, to not include versions and comments, they will be handled by this new API


## Checklist before merge
- [x] Related issue ID is placed at the beginning of PR title in \[brackets\] (can be GH issue or Nu Jira issue)
- [x] Code is cleaned from temporary changes and commented out lines
- [x] Parts of the code that are not easy to understand are documented in the code
- [x] Changes are covered by automated tests
- [ ] Showcase in dev-application.conf added to demonstrate the feature
- [ ] Documentation added or updated
- [ ] Added entry in _Changelog.md_ describing the change from the perspective of a public distribution user
- [ ] Added _MigrationGuide.md_ entry in the appropriate subcategory if introducing a breaking change
- [x] Verify that PR will be squashed during merge

arkadius commented on designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/activities/ScenarioActivityRepository.scala:
```diff
@@ -0,0 +1,81 @@
+import pl.touk.nussknacker.ui.process.repository.activities.ScenarioActivityRepository.ModifyCommentError
+import pl.touk.nussknacker.ui.security.api.LoggedUser
+
+trait ScenarioActivityRepository {
```
1. I see that you directly use the repository to record activity. I already do the service (`newactivity.ActivityService`) which was designed to handle it. We try to follow three layer architecture, IMO services shouldn't call their own repositories directly (they should call services). WDYT?
2. When I see multiple similar addXYZ methods, I have a feeling that it would be better to use the command pattern. Thanks to that we can extract a parent class for similar activities and have some common handlers for similar things. Also, it is easier to distinguish between a next adding method and totally different method like `findActivities`

mgoworko replied:
Ad 1 
I totally agree with 3-layer architecture, but would like to do it in next PR that I've started working on today. In next PR I intended to add handling of scenario activities, that were not persisted before (and were not added in this PR because of huge scope), like incoming and outgoing migrations, or technical updates. So:
- in this PR I replaced old ProcessActions with ScenarioActivities, but in some places I tried to not make too huge changes
- in next PR I would refactor it, including the service layer in between
 
 Would it be ok?
 
Ad 2 
Did you mean command pattern on repository level, or only service level? I would prefer to do it in next PR too, because it is related to the refactor from Ad 1

So I agree on both points, but would like to do this refactor in next PR I already started working on.

arkadius replied:
The command pattern is valuable in any layer - everwhere where there are multiple similar methods. I see that you already did something in this direction in this PR as well. With regards to moving mentioned reactors to the next PR - it is fine for me :+1:

arkadius commented on designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/activities/DbScenarioActivityRepository.scala:
```diff
@@ -0,0 +1,904 @@
+      case ScenarioActivityType.ScenarioCreated =>
+        ScenarioActivity
+          .ScenarioCreated(
+            scenarioId = scenarioIdFromEntity(entity),
```
I didn't analyze it to deep. If you say so, let's stay with current approach for now.

arkadius reviewed (approved):
The parts of the code that I read are fine for me, let's w8 for @lukasz-bigorajski approve as he checked it more extensively..

lukasz-bigorajski commented on designer/server/src/main/scala/db/migration/V1_055__CreateScenarioActivitiesDefinition.scala:
```diff
@@ -0,0 +1,122 @@
+import java.util.UUID
+import scala.concurrent.ExecutionContext.Implicits.global
+
+trait V1_055__CreateScenarioActivitiesDefinition extends SlickMigration with LazyLogging {
```
No, you shouldn't. For me it's just faster. If you use ANSI SQL the query should be the same for HSQL and postgres.

lukasz-bigorajski commented on designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/activities/DbScenarioActivityRepository.scala:
```diff
@@ -0,0 +1,860 @@
+      activityId = activityId,
+      activityDoesNotExistError = ModifyCommentError.ActivityDoesNotExist,
+      validateCurrentValue = validateCommentExists(scenarioId),
+      modify = _.copy(comment = None, lastModifiedByUserName = Some(user.username)),
```
I like this solution

lukasz-bigorajski reviewed (approved):
LGTM! I responded to the questions.

mgoworko commented:
Yesterday I tested this change using real data from prod environments. It turned out, that migration was not fully correct, and existing test did not catch that. Some data, including standalone comments, was missing.

Done after review and approve:
- added 9 tests for migrating each old process action type and standalone comments
- fixed migration, including proper handling of comment prefixes and standalone comments
- change in API contract - finishedAt field made optional in two activities, comment added for "run now" activity

lukasz-bigorajski commented on designer/server/src/main/scala/pl/touk/nussknacker/ui/process/repository/activities/DbScenarioActivityRepository.scala:
```diff
@@ -875,17 +876,45 @@ class DbScenarioActivityRepository(override protected val dbRef: DbRef, clock: C
+      newName <- additionalPropertyFromEntity(entity, "newName")
+    } yield (oldName, newName)
+
+    val legacyCommentPattern = """Rename: \[(.+?)\] -> \[(.+?)\]""".r
```
I wonder whether NU was adding comments in such format or any external system? If it was any external system then this logic should not be here and maybe performed manually in target system.

mgoworko replied:
- Before this PR, NU was adding such comments itself, in `pl.touk.nussknacker.ui.process.repository.ProcessRepository`.
- In this PR it is replaced with adding `ScenarioActivity.ScenarioNameChanged` instead of plain text comment
- But old comments are not fully migrated in the db - because it would require parsing comment during the migration and extracting oldName and newName


Out of 17 types of activities, that are supposed to be handled according to analysis od NU-1772:
- 10 are fully migrated and handled in this PR
- 4 were not saved at all (ScenarioCreated, ChangedProcessingMode, OutgoingMigration, PerformedScheduledExecution), will be added in next PR
- 2 were saved, and are still saved, as plain text comments, not named activities (IncomingMigration, AutomaticUpdate)
- 1 remaining, the ScenarioNameChanged, was added in this PR (but old comments will have to be migrated alongise IncomingMigration and AutomaticUpdate comments in next PR, then this todo will be resolved)

In the next PR, ETA next week, I will add those 4 activities that were not saved until now, and will migrate those 3, that were plain text comments until now.

lukasz-bigorajski commented on designer/server/src/test/scala/db/migration/V1_057__MigrateActionsAndCommentsToScenarioActivities.scala:
```diff
@@ -0,0 +1,379 @@
+
+  private def processEntity(user: String, timestamp: Timestamp) = ProcessEntityData(
+    id = ProcessId(-1L),
+    name = ProcessName("2024_Q3_6917_NETFLIX"),
```
I would anonymize this data.

mgoworko replied:
done

lukasz-bigorajski reviewed (approved):
LGTM. One question.
